### ISSUE TITLE
Alternative to Tiltfile for defining team-level config

### ISSUE THREAD

**Issue by theothertomelliott** [windmilleng/tilt#3295]
We have a number of different Tiltfiles in use within our org, and are having some difficulty ensuring settings like `experimental_telemetry_cmd`, `disable_snapshots` and `set_team` are applied correctly everywhere.

It would be helpful to be able to decouple some of these settings from the "project" level settings. A few possible avenues:
* Environment variables, perhaps even defining an "always-run" Tiltfile
* A `.tiltrc` file
* We have a monorepo, so something like Bazel's WORKSPACE would also be an option

**Comment by maiamcc** [comment]
Interesting! Will bubble this up to product :)

To be clear, how much is your concern consistency vs enforcement? There's potentially an implementation difference between "Want to make sure all the settings get applied the same across a project" and "an engineer shouldn't be able to turn on snapshots by locally editing their Tiltfile".

**Comment by theothertomelliott** [comment]
Consistency is the main concern. If someone creates a brand-new Tiltfile inside our monorepo, I'd like to have snaphots disabled by default.

If someone has the know-how to get around it, it's not the end of the world.

**Comment by dbentley** [comment]
I have a Thought, so I'm going to try to be extra clear.

It seems like this is fundamentally about "per-service Tiltfile configuration should be done in the context of a wider config".

The specific idea is "point the per-service config at the wider config".

I think the arrow should point the other way: "the wider config should point at each per-service config".

I call this the "single entry-point Tiltfile". (To be clear, this is not a good name.)

In this model:
* there's one top-level Tiltfile that does all the setup.
* the args to it decide what subset of services/Tiltfiles to run
* users never tilt up on a different Tiltfile.
  * It's nice that bazel build in a subdirectory and have that context be relevant to your command; we could add a way to do that to Tilt

This would be a refactor for you (and is the opposite of how bazel does it), but I think it has real advantages for Tilt. I'm about to run to a meeting, but can write more or chat about it. (We should have a docs page that describes this better.)

**Comment by theothertomelliott** [comment]
Thanks @dbentley! This is the model we've been trying to maintain thus far. We have a root Tiltfile in our monorepo that points to all other Tiltfiles, with the specific applications selected using args.

The problem we've experienced is that it's very easy for someone to create a "rogue" Tiltfile and run that standalone (especially if they dive into the official Tilt docs rather than our internal intro page). This then bypasses the configuration we're relying on for telemetry, snapshot control, etc.

To the point of Bazel context being relevant to subdirectories, do you mean the "//..." style paths from the workspace? Something like that would also be super useful, as we have a lot of "../../../" imports going on.

**Comment by dbentley** [comment]
Do you have all of them named Tiltfile?

Would it help if the per-service Tiltfiles had a distinct fixed name? Like foo/Setup.tlt and bar/Setupt.tlt? (with .tlt being the extension for the Tiltfile language).

A 3rd kind of file is for libraries. E.g. you might have tilt/k8s.tlt and tilt/mesh.tlt for reusable stuff the different per-service files would load and then use?

**Comment by theothertomelliott** [comment]
We do have our library files named `.tilt`, might tweak that to `.tlt` if the VSCode plugin will recognize that extension :)

I've considered something like an `application.tlt` file for each project, which would help to drive people to the root file as convention.

@jfancher I know you've expressed a preference for independent Tiltfiles, would this model be a major issue with your workflow?

**Comment by jfancher** [comment]
I wouldn't call it a major issue with my usage so far, but I don't see how it's preferable. I can appreciate the desire to set certain settings globally (`.rc` style), but the things listed are Tilt runtime settings, nothing to do with starting or running workloads. I think those are separable concepts.

I'm not clear on exactly what the proposal is here; right now we have a root `service -> Tiltfile` map that just does `include(fileMap[service])` for each service listed in args. Is the suggestion to essentially codify that pattern, or would there be more advanced federation and dependency management involved?

I can give more details about a case I've been thinking about where this approach would pose a practical challenge, but to start out more philosophically: I've yet to be convinced that top-down single entrypoint delegation is the right default choice for a monorepo. We have dozens of teams and hundreds of mostly-independent apps; very few subsets of those apps will ever need to be run together, and almost all local runs are "this single app I'm working on, and whatever dependencies that app needs to function". What are the benefits of coordinating "every potential workload in this repo" in a single central location compared to closer to the unit of development?

Of course I don't want to blow this out of proportion -- we historically have used Edward, which is structured similarly top-down, and it's _fine_. But it has never struck me as ideal.

**Comment by dbentley** [comment]
I don't think the single entrypoint is perfect, nor is it philosophically the One True Way. So I'm not going to try to convince you it is. (FWIW, I really like your point of allowing more coordination the closer you get.)

Right now my opinion is close to "whether the root config is included by the child or vice versa is close to syntactic sugar". As we want to support more of the cross-service patterns, I think it's better to support one style well than trying to support both styles... less well.

I'll see if I can dig up some notes on advantages of the single entrypoint. It would be great to know if there are showstoppers with the single entrypoint approach. So the practical challenge example would be super useful. (Happy to take to a video chat as well)

**Comment by jfancher** [comment]
Okay, this is pretty far afield from Tom's original issue, but since you asked... the following is the sort of scenario I'm trying to think through. It's long and a bit abstract, but hopefully illustrates my mindset.

TL;DR: In a complex project with many configuration options, it seems more tractable to start from a target service and fan out to set up its dependencies, compared with starting a bunch of services separately and hope they add up to the system configuration you want.

---

Say I am working on a project `processor`, a component of a streaming pipeline.

`processor` is a sort of 'hub' service, which connects to:
* _state_ - its main data store
* _config_ - processor configuration
* _resources_ - shared auxiliary resources/configuration
* _input_ - streaming input
* _output_ - streaming output

In a live environment, these are handled by a variety of other services and infrastructure components:
* _state_: `processor-db` database
* _config_: `processor-conf` service
  * using `processor-conf-db` database
* _resources_: `registry` service
  * using `registry-db` database
* _input_: connected via `kafka` to `source` service
  * `source` is connected to `source-db`
* _output_: connected via `kafka` to `sink` service

This is a ton of stuff to run, and it also requires seed data copied or generated to make it do anything useful. Luckily, each of these also has alternatives:

* _state_: may be configured to use an ephemeral in-memory store, **or** connect to the shared dev `processor-db`
* _config_: may be stubbed out to read static config from a local disk, **or** connect to the shared dev `processor-conf-db`
* _resources_: similarly, may be configured statically **or** connect to the shared dev `registry`
* _input_: may be bypassed (data injected directly) **or** connect to a topic in the shared dev `kafka`
* _output_: may be stubbed (write to console) **or** write to the shared dev `kafka`

Of course, if I do choose to run them locally, each of `registry`, `source`, and `sink` may have similar capabilities (including options for additional dependencies I've left out for brevity).

Some high-level observations:
1. Running the "full" `processor` system locally is not a practical default. It's nice to be able to do, but it's too expensive and complex for day-to-day development.
1. However, there are a variety of useful subsets of components that are useful to run together depending on particular area of development (not _every_ possible permutation, but at least a few).
1. The `processor` deployment needs to be configured explicitly based on which components you want to connect it to.
1. Some infrastructure pieces, like `kafka`, may be shared between `processor` and other services (if they're configured to use it).

With that long wind-up, here's an example of what concretely I'd like to do. To debug the "ingestion" component, I want to bring up locally:

* `source` and `source-db` seeded with some initial configuration
* `processor` with in-memory storage, local configuration, and targetting the shared dev `registry`
* `kafka`, to be shared by the two

This would be a bad way to do this:
```console
/repo $ tilt up source source-db processor registry kafka
```

Because:
* It's too much typing and requires knowledge the `processor` developer may not have at top-of-mind
* It's not reusable -- this set of services and configuration means something specific in the context of `processor` development, so it should be encoded in something I can check in
* Most importantly, per 3. above, it wouldn't work anyway -- outside of local development `processor` has no need to run in so many different configurations, so it's not like it has a service discovery mechanism that says "`source` is running, so I should use that". Something about the k8s resources has to be different to accomplish this (command-line flags/env vars via kustomize right now, though arguably not the best way)

Whereas, if I'm focusing only on the `processor` Tiltfile, I can imagine doing something like:
```console
/repo/path/to/processor $ tilt up --debug-ingestion
```

And the Tiltfile like (pseudocode):
````python
load("//infra/path/to/kafka:kafka.tilt", "kafka_cluster")
load("//src/path/to/source:source.tilt", "source_service")

config.define_bool("debug-ingestion")

processor_yaml = "deployment/local"
if cfg.get("debug-ingestion", False):
  kafka_cluster(topics = ["processor-test-topic"])
  source_service(
      static_config = "local/path/to/processor-specific/config",
      produce_topic = "processor-test-topic",
  )
  processor_yaml = "deployment/local-with-source-service"

k8s_yaml(kustomize(processor_yaml))
````

Of course, this would require structuring the Tiltfiles of the other component as an API rather than just "dump the file's text in here and execute it"; I don't mean to imply that'd be the default or even common, but for a few high-value complex cases it could be super useful. Even for the cases where it is just `include()`, it may be conditional or need to happen in a particular order compared with the rest of the file.

Anyway, maybe this kind of thing is not _impossible_ to do with a top-down management, but approaching it from the point of view of "I'm a `processor` developer, I want to set something up for myself and other developers to do some class of work on `processor`", it seems most natural to silo it under `processor`.

In writing this out, it occurred to me that it's sort of reminiscent of a "declarative-vs-imperative" debate -- if things were purely declarative, then maybe it'd be as you say, that things could be defined anywhere and where they get composed is not that important. To frame it in those terms, I think Tilt it lacking the declarative toolbox w.r.t. dependencies, conditionals, orchestration in general to make cases like the above work, when it's fairly simple to imagine (if not to fully implement) imperatively.

**Comment by dbentley** [comment]
Thanks for the concrete example.

I agree totally that it makes sense for processor to have a lot of logic like this, and it should be grouped with processor. I was thinking something like "if you run tilt up from within the processor directory, it should use the parent Tiltfile to setup team-level stuff and then get processed down here". Obviously a lot of extra support is required to make that easy.

My big remaining question is about cross-constellation work. What if you want to run processor together with another project of similar scope? Does processor get data from some other completely different service? How much cooperation/similarity do you want between those? It's those kinds of issues that make me think it's worth a bit of forcing people to play together (i.e. in the same root Tiltfile) early. WDYT?

**Comment by jfancher** [comment]
That's a good question, but not one I have a lot of insight on.

Speaking only for myself, the set of services I listed in my example is about the upper limit of the universe of stuff I'd like to run locally at once -- or rather, the limit of what I think is practical for my situation. Beyond that point:

* Resource constraints become a concern. Even with just a subset of the `processor` case implemented I'm running into limits of what my laptop can handle (I mostly blame docker-desktop for this though, as it's fine outside of k8s).

* The setup / configuration / data burden of initializing a fresh environment becomes overwhelming.

* Other than debugging production issues where the point of failure is unclear (i.e. for everyday development) it's generally faster and easier to only have to run a single microservice or a small related set. This definitely isn't true for all teams or projects -- to reiterate, only speaking for myself here -- but IMO it's usually worth devoting effort to make it more true (vs effort towards making it easier to run all dependencies).

Additionally, we have a pretty ingrained "fallback to shared development sandbox resources" setup, which mitigates the first two points above. It's not perfect, but it provides a simpler alternative than "figure out how to run it locally" for many dependencies.

At a conceptual level, I do really like the idea of having a self-contained, personal instance of any subset of our entire platform, where you just `up service1 service2 ... serviceN` for any group and it all just works together seamlessly (in particular, as opposed to hundreds of devs stomping around the same shared sandbox). However, whenever I consider the practical work that'd need to be done to enable this universally (at the app/infra/config/data level, not just Tilt) it just doesn't seem tractable, or worth it. Hence my focus on the smaller scale.

I'd be happy to be proven wrong though, to see a "mix and match platform" working at a grand scale without sacrificing flexibility or productivity when developing individual components.
